src/cc65/exprdesc.h

```diff
@@ -104,29 +104,58 @@ enum {
     E_LOC_PRIMARY       = 0x0020,       /* Temporary in primary register */
     E_LOC_EXPR          = 0x0040,       /* A location that the primary register points to */
     E_LOC_LITERAL       = 0x0080,       /* Literal in the literal pool */
+    E_LOC_CODE          = 0x0100,       /* C code label location (&&Label) */
 
-    /* Constant location of some sort (only if rval) */
+    /* Immutable location addresses (immutable bases and offsets) */
     E_LOC_CONST         = E_LOC_NONE | E_LOC_ABS | E_LOC_GLOBAL | E_LOC_STATIC |
-                          E_LOC_REGISTER | E_LOC_LITERAL,
+                          E_LOC_REGISTER | E_LOC_LITERAL | E_LOC_CODE,
 
-    /* lvalue/rvalue in C language's sense */
-    E_MASK_RTYPE        = 0x0100,
-    E_RTYPE_RVAL        = 0x0000,
-    E_RTYPE_LVAL        = 0x0100,
-
-    /* Bit-field? */
-    E_BITFIELD          = 0x0200,
+    /* Not-so-immutable location addresses (stack offsets can change) */
+    E_LOC_QUASICONST    = E_LOC_CONST | E_LOC_STACK,
 
-    /* Test */
-    E_NEED_TEST         = 0x0400,       /* Expression needs a test to set cc */
-    E_CC_SET            = 0x0800,       /* Condition codes are set */
+    /* Expression type modifiers */
+    E_BITFIELD          = 0x0200,       /* Expression is a bit-field */
+    E_ADDRESS_OF        = 0x0400,       /* Expression is the address of the lvalue */
 
-    E_HAVE_MARKS        = 0x1000,       /* Code marks are valid */
+    /* lvalue/rvalue in C language's sense */
+    E_MASK_RTYPE        = 0x0800,
+    E_RTYPE_RVAL        = 0x0000,
+    E_RTYPE_LVAL        = 0x0800,
 
-    E_LOADED            = 0x4000,       /* Expression is loaded in primary */
+    /* Expression status */
+    E_LOADED            = 0x1000,       /* Expression is loaded in primary */
+    E_CC_SET            = 0x2000,       /* Condition codes are set */
+    E_HAVE_MARKS        = 0x4000,       /* Code marks are valid */
 
-    E_ADDRESS_OF        = 0x8000,       /* Expression is the address of the lvalue */
+    /* Optimization hints */
+    E_MASK_NEED         = 0x030000,
+    E_NEED_EAX          = 0x000000,     /* Expression needs to be loaded in Primary */
+    E_NEED_NONE         = 0x010000,     /* Expression value is unused */
+    E_NEED_TEST         = 0x020000,     /* Expression needs a test to set cc */
 
+    /* Expression evaluation requirements.
+    ** Usage: (Flags & E_EVAL_<Flag>) == E_EVAL_<Flag>
+    */
+    E_MASK_EVAL         = 0xFC0000,
+    E_EVAL_NONE         = 0x000000,     /* No requirements */
+    E_EVAL_CONST        = 0x040000,     /* Result must be immutable */
+    E_EVAL_COMPILE_TIME = 0x0C0000,     /* Result must be known at compile time */
+    E_EVAL_PURE         = 0x100000,     /* Evaluation must have no side effects */
+    E_EVAL_STATIC       = 0x340000,     /* Evaluation must generate no code */
```

Comment:
I think having `STATIC` and `CONST` names that mean different things from `static` and `const` in C is a bit confusing.

Reply:
There have been different meanings of `const`/`constant` in cc65 for a long time, which I think should be enough to keep people aware of that. And `static` in C mostly refers to object linkage rather than expression traits, except for `_Static_assert`, which exactly indicates the same "no runtime" virtual as used here.

Anyways, do you have suggestions for better names? `E_EVAL_IMMUTABLE` could be  good for `E_EVAL_CONST`, but `E_EVAL_NO_CODE`, `E_EVAL_CODELESS` or `E_EVAL_EMPTY_CODE` for `E_EVAL_STATIC`?

Reply:
I didn't realize, but many/most of these are currently unused.

It would definitely help to expand the comments. "must be immutable" doesn't tell me much. Is `(i++, 0)` immutable? Having good documentation is more important than getting the right names right now. 

`E_EVAL_NO_CODE` seems good for `E_EVAL_STATIC`. I'll try to think about the other names.

Reply:
The comment says "**Result** must be immutable". The result of the expression `(i++, 0)` is always `0`. Thus it is immutable.